hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/solr/SolrUtil.java

```diff
@@ -0,0 +1,497 @@
+    children.stream().forEach(volume -> {
+      updateVolumeSize(volume);
+      rootEntity.setSize(rootEntity.getSize() + volume.getSize());
+    });
+  }
+
+  private void addBucketAndPrefixPath(
+      String[] split, EntityReadAccessHeatMapResponse volumeEntity,
+      long readAccessCount, long keySize) {
+    List<EntityReadAccessHeatMapResponse> bucketEntities =
+        volumeEntity.getChildren();
+    EntityReadAccessHeatMapResponse bucket =
+        new EntityReadAccessHeatMapResponse();
+    bucket.setLabel(split[1]);
+    bucketEntities.add(bucket);
+    bucket.setMinAccessCount(readAccessCount);
+    addPrefixPathInfoToBucket(split, bucket, readAccessCount, keySize);
+  }
+
+  private void addPrefixPathInfoToBucket(
+      String[] split, EntityReadAccessHeatMapResponse bucket,
+      long readAccessCount, long keySize) {
+    List<EntityReadAccessHeatMapResponse> prefixes = bucket.getChildren();
+    updateBucketSize(bucket, keySize);
+    String path = Arrays.stream(split)
+        .skip(2).collect(Collectors.joining("/"));
+    EntityReadAccessHeatMapResponse prefixPathInfo =
+        new EntityReadAccessHeatMapResponse();
+    prefixPathInfo.setLabel(path);
+    prefixPathInfo.setAccessCount(readAccessCount);
+    prefixPathInfo.setSize(keySize);
+    prefixes.add(prefixPathInfo);
+    // This is done for specific ask by UI treemap to render and provide
+    // varying color shades based on varying ranges of access count.
+    updateRootLevelMinMaxAccessCount(readAccessCount);
+  }
+
+  private void updateBucketLevelMinMaxAccessCount(
+      EntityReadAccessHeatMapResponse bucket) {
+    List<EntityReadAccessHeatMapResponse> children =
+        bucket.getChildren();
+    if (children.size() > 0) {
+      bucket.setMinAccessCount(Long.MAX_VALUE);
+    }
+    children.stream().forEach(path -> {
+      long readAccessCount = path.getAccessCount();
+      bucket.setMinAccessCount(
+          path.getAccessCount() < bucket.getMinAccessCount() ? readAccessCount :
+              bucket.getMinAccessCount());
+      bucket.setMaxAccessCount(
+          readAccessCount > bucket.getMaxAccessCount() ? readAccessCount :
+              bucket.getMaxAccessCount());
+    });
+  }
+
+  private void updateRootLevelMinMaxAccessCount(long readAccessCount) {
+    EntityReadAccessHeatMapResponse rootEntity =
+        this.entityReadAccessHeatMapRespRef.get();
+    rootEntity.setMinAccessCount(
+        readAccessCount < rootEntity.getMinAccessCount() ? readAccessCount :
+            rootEntity.getMinAccessCount());
+    rootEntity.setMaxAccessCount(
+        readAccessCount > rootEntity.getMaxAccessCount() ? readAccessCount :
+            rootEntity.getMaxAccessCount());
+  }
+
+  private void updateBucketSize(EntityReadAccessHeatMapResponse bucket,
+                                       long keySize) {
+    bucket.setSize(bucket.getSize() + keySize);
+  }
+
+  public void queryLogs(String path, String entityType, String startDate,
```

Comment:
Should avoid concurrent calls into this this queryLogs. The access to entityReadAccessHeatMapRespRef is not thread safe.

Reply:
> Should avoid concurrent calls into this this queryLogs. The access to entityReadAccessHeatMapRespRef is not thread safe.

will AtomicReference to entityReadAccessHeatMapRespRef not make it thread safe ?

Reply:
AtomicReference of entityReadAccessHeatMapRespRef cannot make it thread safe. Say,

thread1         t1                    t2                  t3                    t4                                     t5 
               queryLogs                    generateHeatMap                                      getEntityReadAccessHeatMapResponse
thread 2 
                                     queryLogs                             generateHeatMap       getEntityReadAccessHeatMapResponse

Then both thread1 and thread2 will have the result of thread2's query result. 

So if entityReadAccessHeatMapRespRef is a resource, then synchronized access should be applied to it. Otherwise, make it as the return value of queryLogs can solve the problem.